<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>create gGraph — gG • gGnome</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="create gGraph — gG"><meta property="og:description" content="Wrapper that instantiates a gGraph object from a variety of different inputs.
Only select parameters combos need to be / should be specified simultaneously."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gGnome</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>create gGraph</h1>
    
    <div class="hidden name"><code>gG.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Wrapper that instantiates a gGraph object from a variety of different inputs.</p>
<p>Only select parameters combos need to be / should be specified simultaneously.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gG</span><span class="op">(</span></span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  breaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  junctions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alignments <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  juncs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prego <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  jabba <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cougar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  weaver <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remixt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rck <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nodes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  edges <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  walks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nodeObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  edgeObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  meta <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>genome</dt>
<dd><p>seqlengths or seqinfo object around which to build an empty gGraph</p></dd>


<dt>breaks</dt>
<dd><p>GRanges around which to build a reference gGraph</p></dd>


<dt>junctions</dt>
<dd><p>BND vcf or bedpe file path, gGraph::Junctions object, or GRangesList specifying pairs of locations to reconnect (can be used in conjunction with breaks)</p></dd>


<dt>alignments</dt>
<dd><p>GRanges with fields $cigar $flag and $qname (e.g. output of bamUtils::read.bam or RSeqLib::BWA specifying alignments to a reference, the graph will then represent the (possibly multimapped) implicit nodes and edges representing an "end to end" walk of all queries lifted to the reference.  Insertions are represented as 0-width nodes and multi-maps yield branches.  The resulting graph can be walked to exhaustively or greedily enumerate all possible linear embeddings of the queries on the reference</p></dd>


<dt>prego</dt>
<dd><p>path to PREGO output file. if specified, instantiates a gGraph object from PREGO output with y.field "cn" specifying copy number</p></dd>


<dt>jabba</dt>
<dd><p>path to JaBbA .rds output file, if specified instantiates a gGraph object from JaBbA output with y.field "cn" specifying copy number</p></dd>


<dt>weaver</dt>
<dd><p>path to Weaver output. if specified, instantiates a gGraph object from Weaver output with y.field "cn" specifying copy number</p></dd>


<dt>remixt</dt>
<dd><p>path to RemiXT output, if specified, instantiates a gGraph object from ReMiXT output with y.field "cn" specifying copy number</p></dd>


<dt>rck</dt>
<dd><p>path to RCK output, if specified, instantiates a gGraph object from RCK output with y.field "cn" specifying total copy number</p></dd>


<dt>nodes</dt>
<dd><p>GRanges of unsigned intervals to be rejoined in gGRaph, used in conjunction with edges argument below</p></dd>


<dt>edges</dt>
<dd><p>data.table with field n1, n2, n1.side, n2.side with each row specifying an edge, i.e. which sides ("left" vs "right") of which integer node indices to connect in the gGRaph</p></dd>


<dt>walks</dt>
<dd><p>GRangesList or gWalk to build a "haplograph" around, see ?haplograph</p></dd>


<dt>nodeObj</dt>
<dd><p>gNode object to create a gGraph around (similar to nodes input above), except generated via the $nodes accessor of an existing gGraph object, used in cojunction with gEdge input to create a new gGnome object from an existing one</p></dd>


<dt>meta</dt>
<dd><p>list of metadata to associate with this gGraph object, for now mostly used to populate gTrack visualization parameters</p></dd>


<dt>edgeeObj</dt>
<dd><p>gEdge object to create a gGraph around (similar to edges input above), except generated via the $edges accessor of an existing gGraph object</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A new gGraph object that is the sum of the component gGraphs</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">mygenome</span> <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">'chr1'</span> <span class="op">=</span> <span class="fl">1e6</span>, <span class="st">'chr2'</span> <span class="op">=</span> <span class="fl">1e7</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">myjuncs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">"delly.final.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"gGnome"</span><span class="op">)</span> <span class="co">## junctions</span></span></span>
<span class="r-in"><span><span class="va">mybreaks</span> <span class="op">=</span> <span class="fu">gr.tile</span><span class="op">(</span><span class="va">mygenome</span>, <span class="fl">10000</span><span class="op">)</span> <span class="co">## 1kb tiles across my genome</span></span></span>
<span class="r-in"><span><span class="va">myweaver</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'weaver'</span>, package<span class="op">=</span><span class="st">'gGnome'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">myremixt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'remixt'</span>, package<span class="op">=</span><span class="st">'gGnome'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">myprego</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'intervalFile.results'</span>, package<span class="op">=</span><span class="st">'gGnome'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">myjab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'jabba.simple.rds'</span>, package<span class="op">=</span><span class="st">"gGnome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>genome <span class="op">=</span> <span class="va">mygenome</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 0 nodes, 0 loose ends (0 terminal and 0 internal), and 0 edges (0 REF and 0 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## create a graph from junctions</span></span></span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>breaks <span class="op">=</span> <span class="va">mybreaks</span>, juncs <span class="op">=</span> <span class="va">myjuncs</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 1387 nodes, 158 loose ends (158 terminal and 0 internal), and 1309 edges (1099 REF and 210 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 1387</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 7 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames          ranges strand |  query.id   tile.id loose.left</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;  &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1         6525379      + |      &lt;NA&gt;      &lt;NA&gt;       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1 6569298-6683446      + |      &lt;NA&gt;      &lt;NA&gt;      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1 6683447-8397763      + |      &lt;NA&gt;      &lt;NA&gt;      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1         8401369      + |      &lt;NA&gt;      &lt;NA&gt;       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     loose.right   node.id  snode.id     index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;logical&gt; &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1       FALSE         1         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2       FALSE         2         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3       FALSE         3         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4       FALSE         4         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 86 sequences from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (1383 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 1309 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id type grl.ix CHROM     POS paramRangeID REF   ALT QUAL FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1  ALT      1     1 6683446         &lt;NA&gt;   C &lt;INV&gt;   NA   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2  ALT      2     1 6683447         &lt;NA&gt;   G &lt;INV&gt;   NA   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3  ALT      3     1 8397763         &lt;NA&gt;   C &lt;INV&gt;   NA   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4  ALT      4     1 8734021         &lt;NA&gt;   G &lt;DUP&gt;   NA   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        CIEND     CIPOS CHR2       END PE MAPQ SR SRQ</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:     -3, 3     -3, 3    1 228054143  7   37 10   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:     -4, 4     -4, 4    1 228054031  6   37 10   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3: -137, 137 -137, 137    1   8401369 14   37 NA  NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:     -7, 7     -7, 7    1   9010792 37   37 10   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                                                       CONSENSUS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                             CGGTAGAGAAAAATAGTAAGCTGGTTTAGAAACTGACGGGAAGCTCCCCTCACCACTGACTTCAAGGCTTCCTCCTTCTCTGTCAGCCGGTTTTAAGGACCCCCGTGGGGCC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:                            GGCAAATTCCTCTCCAATGCGTCCTGGCTGTTTGCCCTCGCTAGTATTGTATGGATGCCTTTCCTCAGTTCACCCCTGGCCCCACGGGGGTCCTTAAAACCGGCTGACAGAGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                                                                                                                                         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: TGCCAGAGTTTGAGACCAACCTGGGCAACTTGGTGAGACTTCATCTCTACAAAAAATCAAAACATTAGCCAGGAGTGGTGGTTGAGGCAACAGTGAGCCATAATTGCACCACTGCACTCCAGCCTGGGCAACAGAATAAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         CE   CT IMPRECISE PRECISE SVTYPE         SVMETHOD INSLEN HOMLEN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: 1.99460 3to3     FALSE    TRUE    BND EMBL.DELLYv0.7.6      0      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: 1.96934 5to5     FALSE    TRUE    BND EMBL.DELLYv0.7.6      0      5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:      NA 3to3      TRUE   FALSE    BND EMBL.DELLYv0.7.6     NA     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: 1.97801 5to3     FALSE    TRUE    BND EMBL.DELLYv0.7.6      0      8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     RDRATIO SOMATIC           MATEID           mateid svtype MATECHROM</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: 0.875842    TRUE expINV00000017:2 expINV00000017:2    BND         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: 0.875842    TRUE expINV00000018:2 expINV00000018:2    BND         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3: 0.874999    TRUE expINV00000020:2 expINV00000020:2    BND         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: 1.680150    TRUE expDUP00000146:2 expDUP00000146:2    BND         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      MATEPOS tier edge.id    class n1.side n2.side n1 n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: 228054143    2       1 INV-like   right   right  2 16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: 228054031    2       2 INV-like    left    left  3 16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:   8401369    2       3 INV-like   right   right  3  4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:   9010792    2       4 DUP-like    left   right  5  5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (1305 additional edges)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## import a genome graph from popular callers</span></span></span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>jabba <span class="op">=</span> <span class="va">myjab</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 895 nodes, 448 loose ends (170 terminal and 278 internal), and 1079 edges (810 REF and 269 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 895</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 16 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames          ranges strand |        cn  start.ix    end.ix eslack.in</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1       1-6525379      + |         4         1        73         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1 6525380-6569292      + |         2        74        76         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1 6569293-6570681      + |         3        77        78         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1 6570682-6570788      + |         4        79        79         0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     eslack.out     loose        edges.in       edges.out   tile.id     index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;numeric&gt; &lt;logical&gt;     &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1          0     FALSE            ()-&gt; -&gt;2(2),-&gt;129(2)         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2          0     FALSE          1(2)-&gt;          -&gt;3(2)         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3          0     FALSE 2(2)-&gt;,899(1)-&gt;          -&gt;4(3)         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4          0     FALSE 3(3)-&gt;,898(1)-&gt; -&gt;5(3),-&gt;814(1)         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      snode.id loose.left loose.right loose.cn.left loose.cn.right   node.id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     &lt;numeric&gt;  &lt;logical&gt;   &lt;logical&gt;     &lt;numeric&gt;      &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1         1       TRUE       FALSE             0              0         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2         2      FALSE       FALSE             0              0         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3         3      FALSE       FALSE             0              0         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4         4      FALSE       FALSE             0              0         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 85 sequences from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (891 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 1079 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id cn type CHROM     POS paramRangeID  REF           ALT QUAL FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1  2  REF  &lt;NA&gt;      NA         &lt;NA&gt; &lt;NA&gt;                 NA   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2  2  ALT     1 6525379         &lt;NA&gt;    C C[3:85232671[   99   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3  2  REF  &lt;NA&gt;      NA         &lt;NA&gt; &lt;NA&gt;                 NA   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4  3  REF  &lt;NA&gt;      NA         &lt;NA&gt; &lt;NA&gt;                 NA   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    DISC_MAPQ NUMPARTS MATEMAPQ MATEID MAPQ REPSEQ READNAMES</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        NA       NA       NA   &lt;NA&gt;   NA   &lt;NA&gt;          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:       255        3       NA 6536:2   60   &lt;NA&gt;          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        NA       NA       NA   &lt;NA&gt;   NA   &lt;NA&gt;          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        NA       NA       NA   &lt;NA&gt;   NA   &lt;NA&gt;          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                        SCTG SVTYPE BX NM MATENM SPAN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                     &lt;NA&gt;   &lt;NA&gt;    NA     NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: c_1_6517001_6542001_163C    BND     0      0   -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                     &lt;NA&gt;   &lt;NA&gt;    NA     NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:                     &lt;NA&gt;   &lt;NA&gt;    NA     NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                                                                                                                        INSERTION</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                                                                                                                                                                                                          &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: ATAGAGCCCAAACCCAATTAATCTTCTAAAAATGGGTAGCCTCTCTTTGACAGCACAGTGCAATTGCATGGTTTTGGAAGCTGAGAGAAAATATTCTATCATGTCTATATTCTTTTGGCCTCTAACATGTATTAAAATTGATATTATTTTAATGATTTCTCAATATTTTGAATCTCTATACATTGAAAGTTTAAAACAATATAAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                                                                                                                                                                                                          &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:                                                                                                                                                                                                          &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    SUBN HOMLEN HOMSEQ IMPRECISE EVDNC SECONDARY mateid svtype first right</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:   NA     NA   &lt;NA&gt;        NA  &lt;NA&gt;        NA   &lt;NA&gt;   &lt;NA&gt;    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:   NA     NA   &lt;NA&gt;     FALSE TSI_G     FALSE 6536:2    BND  TRUE  TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:   NA     NA   &lt;NA&gt;        NA  &lt;NA&gt;        NA   &lt;NA&gt;   &lt;NA&gt;    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:   NA     NA   &lt;NA&gt;        NA  &lt;NA&gt;        NA   &lt;NA&gt;   &lt;NA&gt;    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        coord     mcoord mix MATECHROM  MATEPOS tier edge.id    class n1.side</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:      &lt;NA&gt;       &lt;NA&gt;  NA      &lt;NA&gt;       NA   NA       1      REF   right</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: 1:6525379 3:85232671 122         3 85232670    2       2 TRA-like   right</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:      &lt;NA&gt;       &lt;NA&gt;  NA      &lt;NA&gt;       NA   NA       3      REF   right</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:      &lt;NA&gt;       &lt;NA&gt;  NA      &lt;NA&gt;       NA   NA       4      REF   right</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    n2.side n1  n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:    left  1   2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:    left  1 129</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:    left  2   3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:    left  3   4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (1075 additional edges)</span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>weaver <span class="op">=</span> <span class="va">myweaver</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 2 out-of-bound ranges located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 2 out-of-bound ranges located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 2 out-of-bound ranges located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 2 out-of-bound ranges located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 2 out-of-bound ranges located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>GRanges object contains 1 out-of-bound range located on sequence 17.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Note that ranges located on a sequence whose length is unknown (NA) or</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   on a circular sequence are not considered out-of-bound (use</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   seqlengths() and isCircular() to get the lengths and circularity flags</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   of the underlying sequences). You can use trim() to trim these ranges.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   See ?`trim,GenomicRanges-method` for more information.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 7124 nodes, 6382 loose ends (46 terminal and 6336 internal), and 7427 edges (7101 REF and 326 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 7124</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 8 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames        ranges strand |       acn       bcn        cn loose.left</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;  &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1       1-10000      + |      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1   10001-11000      + |        15         0        15       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1  11001-564368      + |         9         0         9      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1 564369-565368      + |       155         0       155       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     loose.right   node.id  snode.id     index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;logical&gt; &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1       FALSE         1         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        TRUE         2         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3       FALSE         3         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        TRUE         4         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 23 sequences from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (7120 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 7427 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id type grl.ix allele1 allele2 cn unknown1 unknown2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1  ALT      1      -1      -1  1        3     10/0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2  ALT      2      -1      -1  9        3      4/3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3  ALT      3       1       1  5        3    22/10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4  ALT      4      -1      -1  3        3     10/0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    timing    class edge.id n1.side n2.side n1   n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: somatic_post_aneuploid DUP-like       1    left   right 14   24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: somatic_post_aneuploid DEL-like       2   right    left 14   26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3: somatic_post_aneuploid TRA-like       3   right   right 57 4531</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: somatic_post_aneuploid INV-like       4    left    left 60   62</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (7423 additional edges)</span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>prego <span class="op">=</span> <span class="va">myprego</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 604 nodes, 54 loose ends (54 terminal and 0 internal), and 690 edges (576 REF and 114 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 604</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 8 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames          ranges strand |        cn     left.tag    right.tag</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;  &lt;character&gt;  &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1       1-6525379      + |         2       c1_1_1 c1_1_6525379</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1 6525380-6570788      + |         2 c2_1_6525380 c2_1_6570788</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1 6570789-8175806      + |         2 c3_1_6570789 c3_1_8175806</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1 8175807-8401105      + |         2 c4_1_8175807 c4_1_8401105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     loose.left loose.right   node.id  snode.id     index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;logical&gt;   &lt;logical&gt; &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1       TRUE       FALSE         1         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2      FALSE       FALSE         2         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3      FALSE       FALSE         3         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4      FALSE       FALSE         4         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 24 sequences from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (600 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 690 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id        node1 chr1    pos1        node2 chr2    pos2 cn type n1.left</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1 c1_1_6525379    1 6525379 c2_1_6525380    1 6525380  2  REF      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2 c2_1_6570788    1 6570788 c3_1_6570789    1 6570789  2  REF      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3 c3_1_8175806    1 8175806 c4_1_8175807    1 8175807  2  REF      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4 c4_1_8401105    1 8401105 c5_1_8401106    1 8401106  2  REF      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    n1.right n2.left n2.right edge.id class n1.side n2.side n1 n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1       2       NA       1   REF   right    left  1  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2       3       NA       2   REF   right    left  2  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3       4       NA       3   REF   right    left  3  4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4       5       NA       4   REF   right    left  4  5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (686 additional edges)</span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>remixt <span class="op">=</span> <span class="va">myremixt</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 6145 nodes, 94 loose ends (46 terminal and 48 internal), and 6203 edges (6122 REF and 81 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 6145</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 37 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames         ranges strand |     length major_is_allele_a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;         &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1       1-177417      + |  11278.414                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1  177418-267719      + |   2372.545                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1  267720-471368      + |    127.247                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1 471369-1000000      + | 262848.023                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     major_readcount minor_readcount readcount allele_ratio major_depth</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;integer&gt;       &lt;integer&gt; &lt;integer&gt;    &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1              38               2     18510     0.050000    1.559129</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2               0               0      5384     0.000000    2.269293</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3               0               0      2429     0.000000   19.088788</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4           18683            2598     82956     0.122081    0.277075</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     minor_depth total_depth   major_0   minor_0   major_1   minor_1   major_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;numeric&gt;   &lt;numeric&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1   0.0820594    1.641188         1         1         4         2         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2   0.0000000    2.269293         1         1         4         2         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3   0.0000000   19.088788         1         1         4         2         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4   0.0385292    0.315604         1         1         4         2         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       minor_2 major_raw minor_raw major_depth_e minor_depth_e total_depth_e</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1         2   66.8658  3.312029     0.0980487     0.0515662      0.149615</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2         2   97.4220 -0.218737     0.0980487     0.0515662      0.149615</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3         2  821.1136 -0.218737     0.0980487     0.0515662      0.149615</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4         2   11.7030  1.439058     0.0980487     0.0515662      0.149615</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        major_e     minor_e    total_e major_raw_e minor_raw_e major_diff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;  &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1  1105.8340   581.58517  1687.4191           4           2          0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2   232.6250   122.34318   354.9682           4           2          0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3    12.4765     6.56167    19.0381           4           2          0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4 25771.9109 13554.07911 39325.9900           4           2          0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     minor_diff prob_is_outlier_total prob_is_outlier_allele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;integer&gt;             &lt;numeric&gt;              &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1          0                  0.01                   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2          0                  0.01                   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3          0                  0.01                   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4          0                  0.01                   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     total_likelihood_mask allele_likelihood_mask        cn loose.left</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 &lt;integer&gt;              &lt;integer&gt; &lt;integer&gt;  &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1                     0                      0         6       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2                     0                      0         6      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3                     0                      0         6      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4                     0                      0         6      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     loose.right   node.id  snode.id     index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;logical&gt; &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1       FALSE         1         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2       FALSE         2         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3       FALSE         3         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4       FALSE         4         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 23 sequences from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (6141 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 6203 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id type grl.ix prediction_id cn cn_0 cn_1 cn_2 n_1 side_1 n_2 side_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1  ALT      1             2  1    0    1    1 318      1 197      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2  ALT      2             3  1    0    1    1 409      0 456      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3  ALT      3             4  1    0    1    1 430      1 432      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4  ALT      4             5  1    0    1    1 591      0 592      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    edge.id    class n1.side n2.side  n1   n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:       1 TRA-like   right    left  20 3253</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:       2 DUP-like    left   right 111  158</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:       3 DEL-like   right    left 132  134</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:       4 DUP-like    left   right 293  294</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (6199 additional edges)</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## hard code a graph from GRanges and edges</span></span></span>
<span class="r-in"><span><span class="va">nodes</span> <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="fu">GRanges</span><span class="op">(</span><span class="st">"1"</span>,<span class="fu">IRanges</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">100</span><span class="op">)</span>,<span class="st">"*"</span><span class="op">)</span>, <span class="fu">GRanges</span><span class="op">(</span><span class="st">"1"</span>,<span class="fu">IRanges</span><span class="op">(</span><span class="fl">101</span>,<span class="fl">200</span><span class="op">)</span>,<span class="st">"*"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               <span class="fu">GRanges</span><span class="op">(</span><span class="st">"1"</span>,<span class="fu">IRanges</span><span class="op">(</span><span class="fl">201</span>,<span class="fl">300</span><span class="op">)</span>,<span class="st">"*"</span><span class="op">)</span>, <span class="fu">GRanges</span><span class="op">(</span><span class="st">"1"</span>,<span class="fu">IRanges</span><span class="op">(</span><span class="fl">301</span>,<span class="fl">400</span><span class="op">)</span>,<span class="st">"*"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               <span class="fu">GRanges</span><span class="op">(</span><span class="st">"1"</span>,<span class="fu">IRanges</span><span class="op">(</span><span class="fl">401</span>,<span class="fl">500</span><span class="op">)</span>,<span class="st">"*"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">edges</span> <span class="op">=</span> <span class="fu">data.table</span><span class="op">(</span>n1 <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   n2 <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">5</span>,<span class="fl">4</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   n1.side <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   n2.side <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gg</span> <span class="op">=</span> <span class="fu">gG</span><span class="op">(</span>nodes <span class="op">=</span> <span class="va">nodes</span>, edges <span class="op">=</span> <span class="va">edges</span>, meta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gr.colorfield <span class="op">=</span> <span class="st">'type'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## mostly for developer use </span></span></span>
<span class="r-in"><span><span class="fu">gG</span><span class="op">(</span>nodeObj <span class="op">=</span> <span class="va">gg</span><span class="op">$</span><span class="va">nodes</span>, edgeObj <span class="op">=</span> <span class="va">gg</span><span class="op">$</span><span class="va">edges</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gGraph with 5 nodes, 3 loose ends (3 terminal and 0 internal), and 5 edges (1 REF and 4 ALT)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  comprising:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gNode object of length 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 4 ranges and 5 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     seqnames    ranges strand | loose.left loose.right   node.id  snode.id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;logical&gt;   &lt;logical&gt; &lt;integer&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1        1     1-100      + |      FALSE        TRUE         1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2        1   101-200      + |      FALSE       FALSE         2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3        1   201-300      + |      FALSE       FALSE         3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4        1   301-400      + |      FALSE        TRUE         4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (1 additional nodes)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gEdge object with 5 edges</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sedge.id type    class edge.id n1.side n2.side n1 n2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        1  ALT DUP-like       1   right    left  3  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        2  ALT DEL-like       2   right    left  2  4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        3  ALT INV-like       3    left    left  4  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        4  ALT DUP-like       4    left   right  1  5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ... (1 additional edges)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Xiaotong Yao, Marcin Imielinski.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

