<!DOCTYPE html>
<<<<<<< HEAD
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Proximity analysis of two genomic regions and rearrangment movement — proximity • gGnome</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Proximity analysis of two genomic regions and rearrangment movement — proximity"><meta property="og:description" content='Takes a set of n "query" elements (GRanges object, e.g. genes) and determines
their proximity to m "subject" elements (GRanges object, e.g. regulatory
elements) subject to set of rearrangement adjacencies (GRangesList with width 1 range pairs)'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
=======
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>proximity — proximity • gGnome</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="proximity — proximity"><meta property="og:description" content='Takes a set of n "query" elements (GRangse object, e.g. genes) and determines their proximity to m "subject" elements
(GRanges object, e.g. regulatory elements) subject to set of rearrangement adjacencies (GRangesList with width 1 range pairs)'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
>>>>>>> upstream/master
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gGnome</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
<<<<<<< HEAD
    <h1>Proximity analysis of two genomic regions and rearrangment movement</h1>
=======
    <h1>proximity</h1>
>>>>>>> upstream/master
    
    <div class="hidden name"><code>proximity.Rd</code></div>
    </div>

    <div class="ref-description">
<<<<<<< HEAD
    <p>Takes a set of n "query" elements (GRanges object, e.g. genes) and determines
their proximity to m "subject" elements (GRanges object, e.g. regulatory
elements) subject to set of rearrangement adjacencies (GRangesList with width 1 range pairs)</p>
=======
    <p>Takes a set of n "query" elements (GRangse object, e.g. genes) and determines their proximity to m "subject" elements
(GRanges object, e.g. regulatory elements) subject to set of rearrangement adjacencies (GRangesList with width 1 range pairs)</p>
>>>>>>> upstream/master
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">proximity</span><span class="op">(</span></span>
<span>  <span class="va">gg</span>,</span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">subject</span>,</span>
<<<<<<< HEAD
=======
<span>  reduce <span class="op">=</span> <span class="cn">TRUE</span>,</span>
>>>>>>> upstream/master
<span>  ignore.strand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  strict.ref <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chunksize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max.dist <span class="op">=</span> <span class="fl">1e+06</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gg</dt>
<dd><p>gGraph of the "alternate genome"</p></dd>


<dt>query</dt>
<dd><p>GRanges of "intervals of interest" eg regulatory elements</p></dd>


<dt>subject</dt>
<dd><p>GRanges of "intervals of interest" eg genes</p></dd>


<dt>ignore.strand</dt>
<<<<<<< HEAD
<dd><p>whether to ignore strand of input GRanges. Default: True</p></dd>


<dt>verbose</dt>
<dd><p>logical flag, verbose output. Default: False</p></dd>


<dt>mc.cores</dt>
<dd><p>how many cores to use for the path exploration step or if
chunksize is provided, across chunks. Default: 1</p></dd>


<dt>strict.ref</dt>
<dd><p>Boolean, only use actual reference edges in graph.
Default: False</p></dd>


<dt>chunksize</dt>
<dd><p>chunks to split subject and query into to minimize memory
usage, if mc.cores&gt;1 then each chunk will be allotted a core. Default: NULL</p></dd>


<dt>max.dist</dt>
<dd><p>maximum genomic distance to store and compute (1MB by default)
should the maximum distance at which biological interactions may occur.
Default: 1e6</p></dd>
=======
<dd><p>whether to ignore strand of input GRanges</p></dd>


<dt>verbose</dt>
<dd><p>logical flag</p></dd>


<dt>mc.cores</dt>
<dd><p>how many cores to use for the path exploration step or if chunksize is provided, across chunks (default 1)</p></dd>


<dt>chunksize</dt>
<dd><p>chunks to split subject and query into to minimize memory usage, if mc.cores&gt;1 then each chunk will be allotted a core</p></dd>


<dt>max.dist</dt>
<dd><p>maximum genomic distance to store and compute (1MB by default) should the maximum distance at which biological interactions may occur</p></dd>


<dt>ref</dt>
<dd><p>gGraph of the "reference genome", by is the reference genome but can be any gGraph</p></dd>
>>>>>>> upstream/master

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>gWalk object each representing a proximity</p>
    </div>
    <div id="details">
    <h2>Details</h2>
<<<<<<< HEAD
    <p>This analysis makes the (pretty liberal) assumption that all pairs of adjacencies
that can be linked on a gGraph path are in cis (i.e. share a chromosome) in
the tumor genome.</p>
=======
    <p>This analysis makes the (pretty liberal) assumption that all pairs of adjacencies that can be linked on a gGraph path are in
cis (i.e. share a chromosome) in the tumor genome.</p>
>>>>>>> upstream/master
<p>Each output proximity is a gWalk that connects query-subject on the genome
described by gGraph gg.  Each gWalk is  annotated by the metadata of the
corresponding query-subject GRanges pair as well as fields "altdist" and "refdist"
specifying the "alternate and "reference" gGraph distance of the
query-subject pair.  The gWalk metadata field "reldist" specifies the relative
distance (i.e. ratio of altdist to refdist) for that walk.</p>
<<<<<<< HEAD
<p>For more details follow the Proximity Analysis in the gGnome Tutorial:</p>
<p><a href="http://mskilab.com/gGnome/tutorial.html#Proximity_analysis" class="external-link">Proximity Analysis</a></p>
=======
>>>>>>> upstream/master
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Xiaotong Yao, Marcin Imielinski.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

